### ISSUE TITLE
Create a payment workflow

### ISSUE THREAD

**Issue by yoob5** [shibesorg/payments-api#1]
I've started a [payment workflow](https://www.draw.io/#Hshibesorg%2Fpayments-api%2Fmaster%2Fpayment-workflow.xml) that I expect it to be the minimal representation of a front-end UI using the API, e.g. WordPress plugin.

Once we agree on the workflow, we start building the API based on that and start documenting it.

*I moved this conversation from Trello, here's [the card](https://trello.com/c/gx0fNwDG/2-setup-processing-api-project)*

Let me know what do you think.

cc @JamesShelley @PrestackI

**Comment by MoT3rror** [comment]
Questions/Thoughts
- How to handle incorrect amount sent?
- How many temporary addresses can we have?
- How long to wait for cancel payment?

**Comment by yoob5** [comment]
@MoT3rror 

> How to handle incorrect amount sent?

with an incorrect amount or any other error, it should issue a refund. So we need to ask for a refund address like shapeshift.io

> How many temporary addresses can we have?

For each transaction.

> How long to wait for cancel payment?

To be decided.

**Comment by PrestackI** [comment]
I have updated the flowchart with what I think should be included. What do you guys think?

**Comment by yoob5** [comment]
@PrestackI that's great, more explicit. Love it.

**Comment by yoob5** [comment]
> System checks address every X seconds for new transactions

This part should be done on the front-end really, sending a `GET` request to an endpoint, e.g., `/api/v1/<address>/transactions`.

So, I think we should abstract the back-end only functionality from this chart. Or better, start defining the endpoints.

**Comment by MoT3rror** [comment]
I am not a big fan of the front end pinging the backend every X seconds for transactions. Could slow down the back end quick.

I would prefer it use webhooks like Stripe (https://stripe.com/docs/webhooks) or IPN with PayPal. Basically we notify the frontend when the transaction is complete.

**Comment by yoob5** [comment]
@MoT3rror true, let's take a look at how [Shapeshift](https://shapeshift.io/) does it.

**Comment by yoob5** [comment]
@MoT3rror thinking about that, it'll take backend resources to compute listening for transactions and then informing the front-end, doing this per transaction can slow down the backend too.

If we wait and listen for requests and respond might take less effort from the backend.

But, I'm speculating.

**Comment by yoob5** [comment]
We can do either way, using dogechain.info's streaming API https://dogechain.info/api/websocket or doing a listener on the backend as @MoT3rror suggested. The simplest way seems @MoT3rror 's.

**Comment by dayio** [comment]
Hi, 

I think a way to manage the polling of the current transaction status is to add workers for the plugin or the library, something that can runs by http ( long poll, sockets ) or by a process manager like supervirord but we are limited if the direction is a wordpress plugin because most of the wordpress instances are not accessible by ssh or most of the users aren't comfortable with command line.

Pinging the backend can be a problem on the long run beacause for the we must use service workers to call or notify the user and if there is a lot of transactions pending it can be a bit slow.

For the webhook, we must use webhooks if the payment system is out of the website in order to notify the main website that we've done or fail on the transaction

If the system is hosted, we can try to make the payment api as a service of the plugin ( like the good ol days with cgi ) and provide a proxy in php for calling it websockets instead and a fallback as long poll if not supported ( + websockets for dogechain ) 

After it's possible to make this thing in two tier, like a simple plugin for wordpress / any other framework langage and a hostable gateway for checking and make payments if the user can install or use one "cloud" already hosted for making payments / checking
